### PR TITLE
[Android][iOS][notifications] Support unregistering from push notifications

### PR DESCRIPTION
**Pull request by sausti**
# Why
Prior to this change, Expo Notifications did not expose any functionality for unregistering a user from push notifications. This would result in users continuing to receive notifications after logout.

# How
This change adds `unregisterForNotificationsAsync()`, which delegates to the asynchronous [deleteToken()](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/FirebaseMessaging#deleteToken()) method on Android (which will delete a user's FCM token and thus unregister them from FCM) and [unregisterForRemoteNotifications() ](https://developer.apple.com/documentation/uikit/uiapplication/1623093-unregisterforremotenotifications)on iOS.

`unregisterForRemoteNotifications()` on iOS is a synchronous call, but for platform consistency we return a Promise here that will always resolve. This functionality is not applicable to web.

# Test Plan
**Programmatic:** Unit testing added for Android. Unit testing for iOS does not seem beneficial, as this is a simple synchronous call.

**End-to End:** A convenience `"Unregister for push notifications"` function is added to the Bare Expo app's `NotificationsScreen`. One can confirm that this functionality works with the sample app by:
1. Registering their device for push notifications.
2. Sending a test notification to their device and confirming receipt.
3. Unregistering their device for push notifications.
4. Sending a test notification to their device and confirming that it is no longer received.

# Checklist
- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-notifications/ios/EXNotifications/PushToken/EXPushTokenModule.m [hunk 1]
```diff
@@ -45,6 +48,13 @@ @implementation EXPushTokenModule
   });
 }
 
+EX_EXPORT_METHOD_AS(unregisterForNotificationsAsync,
+        resolve:(EXPromiseResolveBlock)resolve reject:(EXPromiseRejectBlock)reject)
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/PushToken/EXPushTokenModule.m:52]:
```suggestion
EX_EXPORT_METHOD_AS(unregisterForNotificationsAsync,
        	    unregisterForNotificationsAsync:(EXPromiseResolveBlock)resolve reject:(EXPromiseRejectBlock)reject)
```

File: packages/expo-notifications/ios/EXNotifications/PushToken/EXPushTokenModule.m [hunk 2]
```diff
@@ -20,6 +20,9 @@ @interface EXPushTokenModule ()
 @property (nonatomic, strong) EXPromiseResolveBlock getDevicePushTokenResolver;
 @property (nonatomic, strong) EXPromiseRejectBlock getDevicePushTokenRejecter;
 
+@property (nonatomic, strong) EXPromiseResolveBlock unregisterForNotificationsResolver;
+@property (nonatomic, strong) EXPromiseRejectBlock unregisterForNotificationsRejecter;
+
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/PushToken/EXPushTokenModule.m:25]: Looks like it's no longer used 😉 

```suggestion
```

File: packages/expo-notifications/CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,7 @@
 - Fixed Android 12+ runtime crash caused by `PendingIntent` misconfiguration. ([#17333](https://github.com/expo/expo/pull/17333) by [@kudo](https://github.com/kudo))
 - Fix app not bringing to foreground when clicking notification on Android 12+. ([#17686](https://github.com/expo/expo/pull/17686) by [@kudo](https://github.com/kudo))
 - Fixed Android data-only FCM notifications (i.e. notifications without a title and message) appearing in the notification drawer ([#17707](https://github.com/expo/expo/pull/17707) by [@sausti](https://github.com/sausti))
+- Add support for un-registering from push notifications on Android and iOS ([#17812](https://github.com/expo/expo/pull/17812) by [@sausti](https://github.com/sausti))
```
Comment (tsapeta) [inline packages/expo-notifications/CHANGELOG.md:16]:
```suggestion
- Add support for unregistering from push notifications on Android and iOS ([#17812](https://github.com/expo/expo/pull/17812) by [@sausti](https://github.com/sausti))
```

**Review by tsapeta — COMMENTED** [review]
Hey, thank you for your contribution! 🙏 
I took the first pass on it, but generally looks good and high quality 💪

File: packages/expo-notifications/ios/EXNotifications/PushToken/EXPushTokenModule.m (hunk 2 above)
Comment (sausti) [reply packages/expo-notifications/ios/EXNotifications/PushToken/EXPushTokenModule.m:25]: Good catch! I forgot to clean this up after changing my approach.

**Review by tsapeta — APPROVED** [review]

**Pull request merged**